### PR TITLE
feat: frequency, min_new_tokens, presence, and repetition penalties

### PR DESCRIPTION
**Pull request by vhain**
Thank you for your contribution, we really appreciate it. The following instructions will help improve your pull request and make it easier to receive feedback. If there are any items you don't understand, don't worry. Just submit the pull request and ask the maintainers for help.

## Motivation

Penalties are defined in [SamplingParams](https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/sampling_params.py#L31-L32), and the tensors being [created](https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/managers/schedule_batch.py#L376-L385) during batch execution, but not yet implemented.

Also we can cut-down some milliseconds in ITL, and some VRAM usage, by not creating those unnecessary tensors.

## Modification

- Implemented following penalties, similarly to they are in [vllm](https://github.com/vllm-project/vllm/blob/66d617e3437e62a6650ffcb85b3190669d37a468/vllm/model_executor/layers/sampler.py#L110-L118):
  - frequency_penalty
  - min_new_tokens
  - presence_penalty
  - repetition_penalty
- Introduces following new `sampling_parameters`, and added them to OpenAI protocol in same name as appeared on vllm for compatibility:
  - `stop_token_ids`: similar to `stop`, but token IDs instead of `str`. This could be useful when used with `min_new_tokens`.
  - `repetition_penalty`
  - `min_new_tokens`
- Each penalizers are fully batch-compatible (filter, merge supported).
- Penalizers will do nothing (not even create a tensor) if they are not required.
- Performance implcations on using penalizer `sampling_parameters` are documented.
- Add ability for `sglang.bench_serving` to accept additional `--request-body` to specify `sampling_params` if desired.

## Benchmark

Benchmarks were ran on single NVIDIA H100 SXM5

Baeline calculated on dc9d06d886151707f97d0b78095df9de262fd3c9:

```
$ python3 -m sglang.bench_serving --backend sglang --port 8413 --dataset-name random --num-prompts 3000 --random-input 256 --random-output 512

============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    inf
Successful requests:                     3000
Benchmark duration (s):                  67.50
Total input tokens:                      378633
Total generated tokens:                  775651
Total generated tokens (retokenized):    774862
Request throughput (req/s):              44.44
Input token throughput (tok/s):          5609.11
Output token throughput (tok/s):         11490.58
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   42141.26
Median E2E Latency (ms):                 45393.52
---------------Time to First Token----------------
Mean TTFT (ms):                          20705.28
Median TTFT (ms):                        14893.99
P99 TTFT (ms):                           49444.93
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          94.00
Median TPOT (ms):                        91.96
P99 TPOT (ms):                           232.66
---------------Inter-token Latency----------------
Mean ITL (ms):                           181.11
Median ITL (ms):                         62.46
P99 ITL (ms):                            441.66
==================================================
```

Calculated on this PR's branch (without any penalizers):

```
$ python3 -m sglang.bench_serving --backend sglang --port 8413 --dataset-name random --num-prompts 3000 --random-input 256 --random-output 512

============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    inf
Successful requests:                     3000
Benchmark duration (s):                  66.11
Total input tokens:                      378633
Total generated tokens:                  775651
Total generated tokens (retokenized):    775118
Request throughput (req/s):              45.38
Input token throughput (tok/s):          5727.04
Output token throughput (tok/s):         11732.16
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   40881.94
Median E2E Latency (ms):                 43967.10
---------------Time to First Token----------------
Mean TTFT (ms):                          19884.75
Median TTFT (ms):                        14226.56
P99 TTFT (ms):                           47738.97
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          91.96
Median TPOT (ms):                        90.11
P99 TPOT (ms):                           308.54
---------------Inter-token Latency----------------
Mean ITL (ms):                           174.54
Median ITL (ms):                         58.56
P99 ITL (ms):                            440.18
==================================================
```

Additional benchmarks can be found in the [updated sampling_params.md](https://github.com/tmfi-us/sglang/blob/ryan/srt/schedule_batch/implement-penalties/docs/en/sampling_params.md#benchmarks)

## Notes to Reviewers

This is my very first time writing Python code, and thus working with PyTorch (my industry backgrounds are primarily Golang, Rust, Java, Nodejs, ...). So please feel to add any comments if something looks off or inefficient.

## Checklist

- [x] 1. Ensure pre-commit `pre-commit run --all-files` or other linting tools are used to fix potential lint issues.
- [x] 2. Confirm that modifications are covered by complete unit tests. If not, please add more unit tests for correctness.
- [x] 3. Modify documentation as needed, such as docstrings or example tutorials.

### CODE REVIEW

**Comment by zhyncs** [comment]
@vhain Nice work!

File: docs/en/sampling_params.md [hunk 1]
```diff
@@ -142,3 +165,252 @@ print(response.json())
 
 The `image_data` can be a file name, a URL, or a base64 encoded string. See also `python/sglang/srt/utils.py:load_image`.
 Streaming is supported in a similar manner as [above](#streaming).
+
+## Performance Implications on Penalties
```
Comment (merrymercy) [inline docs/en/sampling_params.md:170]: This is very helpful. Thanks!

File: python/sglang/bench_serving.py [hunk 2]
```diff
@@ -47,6 +47,7 @@ class RequestFuncInput:
     prompt_len: int
     output_len: int
     model: str
+    request_body: Dict[str, Any]
```
Comment (merrymercy) [inline python/sglang/bench_serving.py]:
```suggestion
    extra_request_body: Dict[str, Any]
```
similar to this https://github.com/sgl-project/sglang/blob/1ac304eeb483c4ce3435dd1673426ddd7271d02c/test/srt/test_openai_server.py#L376

File: test/srt/sampling/penaltylib/penalizers/test_min_new_tokens.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
+import typing
+import unittest
+
+import torch
+
+from sglang.srt.sampling.penaltylib.penalizers.min_new_tokens import (
+    BatchedMinNewTokensPenalizer,
+)
+
+from ..utils import (
```
Comment (merrymercy) [inline test/srt/sampling/penaltylib/penalizers/test_min_new_tokens.py]: I prefer not to use relative imports. In sglang, we avoid them by moving common parts to `python/sglang/test/`

**Review by merrymercy — CHANGES_REQUESTED** [review]
Thanks for your contributions! The code quality is very high and it is well tested. Great work! I only have few minor style comments.

File: test/srt/sampling/penaltylib/test_srt_endpoint_with_penalizers.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+from sglang.srt.utils import kill_child_process
+from sglang.test.test_utils import DEFAULT_MODEL_NAME_FOR_TEST, popen_launch_server
+
+
+class TestBatchPenalizerE2E(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        cls.model = DEFAULT_MODEL_NAME_FOR_TEST
+        cls.base_url = f"http://localhost:{8157}"
```
Comment (zhyncs) [inline test/srt/sampling/penaltylib/test_srt_endpoint_with_penalizers.py]: I don't recommend using localhost because it can cause issues in certain environments.
```
ERROR:    [Errno 99] error while attempting to bind on address ('::1', 8157, 0, 0): cannot assign requested address
```

Write it simply.
```suggestion
        cls.base_url = "http://127.0.0.1:8157"
```

File: docs/en/sampling_params.md [hunk 5]
```diff
@@ -142,3 +165,252 @@ print(response.json())
...
+
+### Memory
+
+Since we compute penalty algorithms through CUDA, the logic stores relevant parameters on GPU. This is usually in a scale of `vocab_size` multiplied by `running_requests`.
+
+You can run your own benchmark with desired parameters on your own hardware to make sure it's not OOMing before using.
+
+Tuning `--mem-fraction-static` and/or `--max-running-requests` will help. See [here](hyperparameter_tuning.md#minor-tune---max-prefill-tokens---mem-fraction-static---max-running-requests) for more information.
+
+### Benchmarks
```
Comment (zhyncs) [inline docs/en/sampling_params.md:188]: I wouldn't strongly recommend including the benchmark results in the document, but if necessary, consider using a collapsible section so users can expand it when needed.

```
<details>
<summary>title</summary>
</details>
```

File: docs/en/sampling_params.md [hunk 6]
```diff
@@ -52,6 +55,26 @@ spaces_between_special_tokens: bool = True,
 regex: Optional[str] = None,
 # Do parallel sampling and return `n` outputs.
 n: int = 1,
+
+## Penalties. See [Performance Implications on Penalties] section below for more informations.
+
+# Float that penalizes new tokens based on their frequency in the generated text so far.
+# Values > 0 encourage the model to use new tokens, while values < 0 encourage the model to
+# repeat tokens.
+frequency_penalty: float = 0.0,
```
Comment (zhyncs) [inline docs/en/sampling_params.md:64]: https://platform.openai.com/docs/api-reference/chat/create#chat-create-frequency_penalty

File: docs/en/sampling_params.md [hunk 7]
```diff
@@ -52,6 +55,26 @@ spaces_between_special_tokens: bool = True,
...
+# Float that penalizes new tokens based on whether they appear in the generated text so far.
+# Values > 0 encourage the model to use new tokens, while values < 0 encourage the model to repeat
+# tokens.
+presence_penalty: float = 0.0,
```
Comment (zhyncs) [inline docs/en/sampling_params.md:68]: https://platform.openai.com/docs/api-reference/chat/create#chat-create-presence_penalty

File: docs/en/sampling_params.md (hunk 6 above)
Comment (zhyncs) [reply docs/en/sampling_params.md:64]: The official OpenAI documentation supports a range from -2 to 2, each value having its specific usage and meaning.
> Reasonable values for the penalty coefficients are around 0.1 to 1 if the aim is to just reduce repetitive samples somewhat. If the aim is to strongly suppress repetition, then one can increase the coefficients up to 2, but this can noticeably degrade the quality of samples. Negative values can be used to increase the likelihood of repetition.

Should we consider aligning with OpenAI?

File: docs/en/sampling_params.md (hunk 7 above)
Comment (zhyncs) [reply docs/en/sampling_params.md:68]: Same as mentioned above.

File: docs/en/sampling_params.md [hunk 8]
```diff
@@ -52,6 +55,26 @@ spaces_between_special_tokens: bool = True,
...
+# Float that penalizes new tokens based on whether they appear in the prompt and the generated text
+# so far. Values > 1 encourage the model to use new tokens, while values < 1 encourage the model to
+# repeat tokens.
+repetition_penalty: float = 1.0,
```
Comment (zhyncs) [inline docs/en/sampling_params.md:72]: If there's already `frequency_penalty` and `presence_penalty`, is there still a need for `repetition_penalty`? Can the first two not address certain situations?

File: python/sglang/srt/managers/schedule_batch.py [hunk 9]
```diff
@@ -390,15 +395,17 @@ def batch_sampling_params(self, vocab_size, int_token_logit_bias):
...
-        self.frequency_penalties = torch.tensor(
-            [r.sampling_params.frequency_penalty for r in reqs],
-            dtype=torch.float,
-            device=device,
-        )
-        self.presence_penalties = torch.tensor(
-            [r.sampling_params.presence_penalty for r in reqs],
-            dtype=torch.float,
+
+        self.penalizer_orchestrator = penaltylib.BatchedPenalizerOrchestrator(
```
Comment (zhyncs) [inline python/sglang/srt/managers/schedule_batch.py:406]: If the user does not **explicitly** want to enable these parameters, is it still necessary to apply them here?

File: docs/en/sampling_params.md (hunk 5 above)
Comment (vhain) [reply docs/en/sampling_params.md:188]: It is already wrapped with collapsible section!

File: docs/en/sampling_params.md (hunk 6 above)
Comment (vhain) [reply docs/en/sampling_params.md:64]: They are already enforced to be in the range, same as OpenAI spec. I will mention about that enforcement to this doc as well. I just mostly copy-pasted from [what it is in vLLM](https://github.com/vllm-project/vllm/blob/main/vllm/sampling_params.py#L49-L60)

File: docs/en/sampling_params.md (hunk 8 above)
Comment (vhain) [reply docs/en/sampling_params.md:72]: When I was implementing the penalty logic, I wrote based on [what it is in vLLM docs](https://github.com/vllm-project/vllm/blob/main/vllm/sampling_params.py#L49-L60), and [how they implemented](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/sampler.py#L201-L268).

They have `frequency_penalty`, `presence_penalty`, and `repetition_penalty` all in place.

`frequency_penalty` is to penalize based on frequency in the output tokens.
`presence_penalty` is to penalize based on the presence (1 or 0) in the output token.

`repetition_penalty` is based on the presence (1 or 0) in the input + output tokens. So it is slightly different than `presence_penalty`.

I would personally want to implement additional penalty for frequency in the input + output token as well. While it is easy to do, I had hard time naming the penalty, and also vLLM does not have this, so I just skipped it for now.

File: python/sglang/srt/managers/schedule_batch.py (hunk 9 above)
Comment (vhain) [reply python/sglang/srt/managers/schedule_batch.py:406]: `BatchedPenalizerOrchestrator` will not do anything unless there is a `Req` with penalties defined.

`BatchedPenalizerOrchestrator` class, when being created, will check if there is any `Req` in the batch with penalty defined, and do the work only if there is any.

Of course we can wrap this with `if` clause to even prevent this class from being initialized, but I think that would just add complexity to the `schedule_batch.py` code while not giving us meaningful performance upside.

File: docs/en/sampling_params.md (hunk 8 above)
Comment (zhyncs) [reply docs/en/sampling_params.md:72]: ok

File: docs/en/sampling_params.md (hunk 6 above)
Comment (zhyncs) [reply docs/en/sampling_params.md:64]: We may include some negative value test cases in the unit tests.

File: python/sglang/srt/managers/schedule_batch.py (hunk 9 above)
Comment (zhyncs) [reply python/sglang/srt/managers/schedule_batch.py:406]: make sense

Comment (vhain) [reply python/sglang/srt/managers/schedule_batch.py:406]: Thanks for pointing out a good one. I will mention in the comment about how `BatchedPenalizerOrchestrator` will behave briefly right above this line.

**Review by merrymercy — APPROVED** [review]

**Comment by Ying1123** [comment]
@vhain ready to merge?
@zhyncs can you take another look and approve this?

**Comment by Ying1123** [comment]
In the future, we plan to wrap all the logits processing and sampling into the Cuda graph as well.
Are these penalizers cuda-graph compatible?

**Review by zhyncs — APPROVED** [review]
overall LGTM

**Comment by zhyncs** [comment]
> In the future, we plan to wrap all the logits processing and sampling into the Cuda graph as well. Are these penalizers cuda-graph compatible?

@vhain

**Comment by vhain** [comment]
@Ying1123 @zhyncs I'm still pretty new to this ground (started playing with LLM ~ month ago, started using `sglang` ~ 2 weeks ago, and this is my first ever Python/PyTorch code written). I don't think this penalizers are implemented based on CUDA grapu, but I will take a look on other CUDA graph implementations in `sglang` to figure it out. Hopefully I can pull up another PR to make it use CUDA graphs...

**Comment by vhain** [comment]
@Ying1123 @zhyncs Although I might need some extra time on what and how the CUDA graph works, all those penalizers are purely written in PyTorch (with `device="cuda"`). So it should be fairly simple to migrate..?

**Comment by merrymercy** [comment]
Yes, do not worry about it for now. We will refactor and start moving the sampling code to be within the CUDA graph. We will let you know later and ask for your help if needed.

**Comment by Xu-Chen** [comment]
We found that presence_penalty has a great influence on the speed of decode.
When `set presence_penalty = 0.5` qwen-72b-gptq-int4 ( ~17 token/s on 4 * A100)，3.x slower than `set presence_penalty=0.0` (~60 token/s on 4*A100).
cc @merrymercy @zhyncs 

This can be reproduced in version 0.3.0, but does not occur in version 0.2.13.

**Comment by merrymercy** [comment]
@hnyls2002 Is this related to how the cuda graph is triggered?

**Comment by merrymercy** [comment]
@Xu-Chen Recently, we did some modifications to how the cuda graph is enabled. In 0.3.0, if you set penalty or regex, cuda graph will be disabled. This is a bug and we will fix it soon.

**Comment by hnyls2002** [comment]
@vhain @Xu-Chen  I believe the performance slowdown problem will be fixed in #1392. 

With llama3.1-8b on 1 A100, the decoding speed is 77 tokens per sec the same as without presence penalty, but on main the speed is 71 tokens per sec.

On 4 x A100, the decoding speed is 129 tokens per sec, while without presence penalty the speed is 134 tokens per sec, but on main the speed is 58 tokens per sec.

**Comment by hnyls2002** [comment]
@vhain @Xu-Chen @merrymercy 
The penalties consume calculation in two phases:
- Cumulate output tokens or input tokens.
- Apply the penalties to the logits results.

The second phase was fully overlapped by https://github.com/sgl-project/sglang/blob/184e4446529bd2518361e62fe2cd93a713c78438/python/sglang/srt/model_executor/model_runner.py#L568 (this calculation can be executed parallel with model forward)

Also, the first phase can be overlapped, I have made some tries, but it failed.

https://github.com/sgl-project/sglang/commit/dc1672596f9c6451573c99a8534695c9f9f72d87
https://github.com/sgl-project/sglang/commit/3273d3fac21d40c87382a58ed6ed5a5b1c7667d7

What I thought was to make the cumulate a lazy operation, only being done when batch size changes or before applying. This can save more time and make the inference with penalties as fast as in normal cases. If you are interested in this, you are welcome to contribute to this optimization.
